lightning/src/ln/channel.rs

```diff
@@ -3276,44 +3376,120 @@ impl<Signer: Sign> Channel<Signer> {
+		let proposed_max_total_fee_satoshis = if self.is_outbound() {
+				proposed_max_feerate as u64 * tx_weight / 1000 + self.config.force_close_avoidance_max_fee_satoshis
+			} else {
+				u64::max_value()
+			};
+
+		self.closing_fee_limits = Some((proposed_total_fee_satoshis, proposed_max_total_fee_satoshis));
+	}
+
+	/// Returns true if we're ready to commence the closing_signed negotiation phase. This is true
+	/// after both sides have exchanged a `shutdown` message and all HTLCs have been drained. At
+	/// this point if we're the funder we should send the initial closing_signed, and in any case
+	/// shutdown should complete within a reasonable timeframe.
+	fn closing_negotiation_ready(&self) -> bool {
+		self.pending_inbound_htlcs.is_empty() && self.pending_outbound_htlcs.is_empty() &&
+			self.channel_state &
+				(BOTH_SIDES_SHUTDOWN_MASK | ChannelState::AwaitingRemoteRevoke as u32 |
+				 ChannelState::PeerDisconnected as u32 | ChannelState::MonitorUpdateFailed as u32)
+				== BOTH_SIDES_SHUTDOWN_MASK &&
+			self.pending_update_fee.is_none()
+	}
+
+	/// Checks if the closing_signed negotiation is making appropriate progress, possibly returning
+	/// an Err if no progress is being made and the channel should be force-closed instead.
+	/// Should be called on a one-minute timer.
+	pub fn timer_check_closing_negotiation_progress(&mut self) -> Result<(), ChannelError> {
+		if self.closing_negotiation_ready() {
+			if self.closing_signed_in_flight {
+				return Err(ChannelError::Close("closing_signed negotiation failed to finish within two timer ticks".to_owned()));
+			} else {
+				self.closing_signed_in_flight = true;
+			}
+		}
+		Ok(())
+	}
+
+	pub fn maybe_propose_first_closing_signed<F: Deref, L: Deref>(&mut self, fee_estimator: &F, logger: &L)
+		-> Result<(Option<msgs::ClosingSigned>, Option<Transaction>), ChannelError>
+		where F::Target: FeeEstimator, L::Target: Logger
+	{
+		if self.last_sent_closing_fee.is_some() || !self.closing_negotiation_ready() {
+			return Ok((None, None));
+		}
+
+		if !self.is_outbound() {
+			if let Some(msg) = &self.pending_counterparty_closing_signed.take() {
+				return self.closing_signed(fee_estimator, &msg);
+			}
+			return Ok((None, None));
+		}
+
+		self.calculate_closing_feerate_limits(fee_estimator);
+		let (our_min_fee, our_max_fee) = self.closing_fee_limits.unwrap();
+
+		assert!(self.shutdown_scriptpubkey.is_some());
+		let (closing_tx, total_fee_satoshis) = self.build_closing_transaction(our_min_fee, false);
+		log_trace!(logger, "Proposing initial closing_signed for our counterparty with a fee range of {}-{} sat (with initial proposal {} sats)",
+			our_min_fee, our_max_fee, total_fee_satoshis);
 
-		let (closing_tx, total_fee_satoshis) = self.build_closing_transaction(proposed_total_fee_satoshis, false);
 		let sig = self.holder_signer
 			.sign_closing_transaction(&closing_tx, &self.secp_ctx)
-			.ok();
-		assert!(closing_tx.get_weight() as u64 <= tx_weight);
```

Comment:
Looks like we got rid of this assert, do we still want that?

Reply:
Hmm, I mean I don't think we need it unless its easy, and its no longer entirely trivial given we calculate the weight and absolute fees in `calculate_closing_fee_limits` potentially long before we get here.